### DISCUSSION TITLE
Поисковик по толстым книгам ВП СССР

### DISCUSSION THREAD

**Discussion by iprst** [terratensor/kob-library-app#16] [Announcements]
Товарищи!
Мы запустили поисковик по текстам толстых книг ВП СССР.

★ https://kob.svodd.ru

Поиск реализуется внутри содержания отдельного параграфа, включая сноски, если они есть. Таким образом текстовый запрос в строке поиска служит цели найти наиболее подходящие под этот запрос параграфы. Рядом с кнопкой «поиск» есть кнопка опций, где можно выбрать режим и подключить «концептуальный словарь синонимов». Режимы поиска позволяют выбрать точное совпадение фразы в результатах (по соответствию фразе), появление хотя бы одного из слов запроса в параграфе (по совпадению слов), или найдя подходящий параграф, можно выбрать его номер и взяв соседние (отняв или добавив единицу), показать несколько параграфов подряд. Это удобно, если нужна более развёрнутая информация, но нет необходимости обращаться к целой книге.

Если цель поиска — найти книгу по искомому запросу, то для этой цели в выдаче к каждой цитате присоединяется название толстой книги, откуда она была получена. Поисковый запрос, к которому вы часто обращаетесь, можно сохранить при помощи кнопки «короткая ссылка» — в ней также сохраняется режим конкретного поиска. Режим «словаря концептуальных терминов» позволяет искать все синонимы (иногда антонимы) понятий, перечисленных в поисковой строке, так, при поиске «пфу» со включенным «концептуальным словарём» будут получены все связанные термины, например «полная функция управления» или «цели», однако если вы поищите «пфу» без словаря, результатом будет лишь единственный параграф, где понятие полной функции управления было обозначено аббревиатурой.

Пример одного из сценариев работы с поисковиком.
Представьте, что вам нужно найти в толстых книгах пару терминов, между которыми вы предполагаете существование какой-либо связи. Если вы воспользуетесь обычным поиском операционной системы или редактора файлов по тексту книги, вы можете найти книгу, содержащую искомые понятия, но это не означает, что в книге эти понятия будут связаны в рамках одной страницы текста или даже одной главы. Понятия просто могут присутствовать в совершенно раздельных частях текста, поскольку тексты толстых книг достаточно обширны и включают много понятий. Такие запросы практически бесполезны при использовании стандартных средств поиска по тексту книги в файле, поскольку в большинстве случаев стандартный поиск слеп к расстоянию между поисковыми словами. В нашем поисковике запрос ограничивается параграфами книг, поэтому поиск связи между понятиями намного эффективнее.

Вот примеры связок, которые будет сложно найти быстро без нашего поисковика:
«предиктор + пфу»
★ https://svodd.ru/exxyAS9p

«достоевский + магия»
★ https://svodd.ru/NYxu2lGg

«пушкин + масоны»
★ https://svodd.ru/hOJ0iS0u

Вопросы, предложения, замечания и найденные ошибки можно публиковать в этом обсуждении.

**Comment by BlackPatrick** [comment]
Здравствуйте, товарищи!
У меня следующий вопрос: а к какой библиотеке обращается поисковик? Это какой-то внешний ресурс или вы собрали книги на svodd.ru? Тоже самое касается и военно-исторической библиотеки? Что это за библиотека?

Спрашиваю потому как имею предложение организовать доступ пользователей к этим библиотекам. Доступ можно реализовать как минимум в двух вариантах. Первый путем открытия книги прямо на сайте, а второй с возможностью скачать ее в каком-нибудь из форматов?

**Reply by iprst** [reply]
Обе библиотеки размещены в базе данных на сервере в форме списка параграфов (по аналогии с комментариями сайта). Источником для ВП СССР служили очищенные от форматирования опубликованные толстые книги, которые парсер побил на параграфы и поместил в базу данных. Источником для военно-исторической библиотеки служили опубликованные в открытом доступе книги легальной библиотеки [милитера](https://militera.lib.ru/). Все эти книги были распознаны в текст, затем парсер пробежался по ним и разбил их на параграфы, разместив их в базу данных. 

Обе библиотеки в обозримом будушем получат новый движок, парсер разобьёт тексты на более удобные фрагменты, поисковые запросы будут полнее, будут проведены некоторые другие улучшения, в том числе выдача контекста в форме веб-странцы, с пролистыванием на нужную страницу и прокруткой экрана до фрагмента. Эти технологии отрабатываются на другой библиотеке, которую в открытый доступ мы не размещали. 

В личном кабинете можно будет сохранять закладки, избранное и тд. 

Все книги доступны в сети, но ни на какие книги у нас нет прав на их раздачу как файла, только цитирование, поэтому скачивание это спорный функционал в настоящий момент. Но для каждой книги приводится её корректное название, и абсолютно все книги можно найти в сети в форме файла.

**Reply by audetv** [reply]
Тексты библиотеки КОБ размещены здесь https://github.com/terratensor/book-parser/tree/main/books/VPSSSR/TXT 
Из этих файлов состоит поисковая база. Мы обрабатывали текстовые версии фалов. 
В итоге поиск производиться не по файлам текстов, а по специальным документам записанным в базу данных, которые составляют поисковый индекс.

Такие документы создает программа-парсер, которая проходит по тексту, разбивает текст на части — параграфы. Разбиение производиться по символу переноса строки, и присоединяет к этими частям-параграфам тексты сносок (примечаний), чтобы они были рядом в поиске. 

Эта самая первая схема, позже мы её улучшили, и в военно-исторической библиотеке блоки — параграфы уже формируются по иному принципу, имеют ограничения по размеру символов 1800 - 3200 символов. Мелкие блоки-предложения собираются в укрупненные блоки (чанки), большие портянки разбиваются на части, все это формируется в блоки ~ по 3200 символов. В общем тексты книг разбиваются на примерно одинаковые блоки, что исключает из поисковой выдачи большие «портянки», или блоки из одного предложения в несколько слов.

Индекс библиотеки КОБ надо переделать, применить обновленный метод, пересобрать ее, переписать парсер-программу, чтобы он присоединил сноски, т.к. парсер общей библиотеки (Военно-исторической и библиотеки локальной), не умеет работать со сносками. Сделаем. 


> скачивание это спорный функционал в настоящий момент.

Да, по поводу файлов, мы специально не предлагаем файлы книг, только поиск по книгам. Но указываем автора и наименование книги, чтобы можно было найти нужную книгу и приобрести или скачать.

Наверное для книг ВП СССР можно применить то, что вы описываете, т.к. эти книги разрешено распространять и копировать и кажется, что они не обременены запретами авторского права. Поправьте меня, если ни так. Но для иных книг это ни так.. 

Есть идея для книг добавить неограниченный контекст, точнее ограниченный размером книги и дать возможность пользователю просматривать собранную обратно в книгу сущность, которая не будет исходным файлом книги, а будет html представлением набором страниц с возможность читать, пролистывать и прочее. Но всё же это не читалка книг в привычном некоторым пользователям формате и не книга. Так же фильтровать книги по автору, наименованию и жанру.

**Reply by iprst** [reply]
> Из этих файлов состоит поисковая база. Мы обрабатывали текстовые версии фалов.

Важно что кобовские книги в техническом архиве сильно отформатированы под задачи поиска, они не сохраняют оформления и сносок, всё оформлено в одномерное представление, то есть формально это не книги ВП СССР, а корпус текста для разметки.

**Reply by iprst** [reply]
> Наверное для книг ВП СССР можно применить

Оригиналы книг хранятся на кобовских сайтах, то есть потребовалось бы либо скопировать и распространять файлы с сервера, либо делать ссылку на чужой сервер, где книги выложены поштутчно. И то и другое пока кажется спорным решением в нашем случае. 

С точки зрения методологии работы с информацией, поисковик никак не пересекается с привычной библиотекой, и это не баг, а фича. Задача текстового поиска это предоставить метод работы с информацией, который исключает-дополняет логику «прочесть целую книгу, чтобы найти этот модуль» — этот модуль отныне и вовекивеков можно найти без прочтения книги, не запрещая её прочтение. Книги необходимо читать. Это отдельный, и не связанный с методологией поиска процесс. Необходимо читать книги. Точка. С информацией необходимо работать и другими методами. Другая точка. 

Методы не исключают друг друга, но между их взаимодействием есть конкретная тенденция: прочтение книги это совершенно другой процесс, в предельном понимании противоположный процессу работы с информацией через текстовый поиск. Это два режима работы, которые имеет смысл разделять, потому что они проводятся в совершенно разных условиях и с совершенно разными предпосылками. 

Читать книги это один метод. Мы предоставили другой метод работы с информацией. У связи между этими методами должна быть очень хорошая причина, а пока наблюдается очень хорошая причина для максимального разделения двух этих методов.
